File: backend/checkpoint.js
/**
 * CHECKPOINT MODULE
 * Pure checkpoint checking logic.
 * Check after every 3 answered questions (valid + invalid count, skipped do not count).
 */

const CHECKPOINTS = [
    { level: 1, points: 280, minQuestions: 10, label: 'Basic' },
    { level: 2, points: 420, minQuestions: 15, label: 'Good' },
    { level: 3, points: 560, minQuestions: 20, label: 'Very Detailed' },
    { level: 4, points: 700, minQuestions: 25, label: 'Maximum Depth' }
];

/**
 * Check if a checkpoint has been reached.
 * Only checks every 3 answered questions.
 * Only checks for the NEXT checkpoint not yet reached.
 * @param {Object} session - The session object
 * @returns {Object|null} Checkpoint info or null
 */
function checkForCheckpoint(session) {
    const answered = session.counters.questionsAnswered;
    const totalPoints = session.points.total;
    const reached = session.checkpoints.reached;

    // Only run checkpoint logic after every 3 ANSWERED questions
    // (skipped do NOT count, invalid DO count) — matches spec Section 7.
    if (answered < 3 || answered % 3 !== 0) {
        return null;
    }

    // Find the next checkpoint not yet reached
    const nextCP = CHECKPOINTS.find(cp => cp.level === reached + 1);
    if (!nextCP) return null; // all checkpoints already reached

    // Check if thresholds are met
    if (totalPoints >= nextCP.points && answered >= nextCP.minQuestions) {
        session.checkpoints.reached = nextCP.level;
        return {
            level: nextCP.level,
            label: nextCP.label,
            questionsAnswered: answered,
            totalPoints: totalPoints,
            autoShow: nextCP.level === 4 // CP4 auto-shows results
        };
    }

    return null;
}

/**
 * Get profile quality label based on checkpoint reached
 */
function getProfileLabel(checkpointReached) {
    const cp = CHECKPOINTS.find(c => c.level === checkpointReached);
    if (cp) return cp.label;
    return 'Preliminary';
}

module.exports = {
    checkForCheckpoint,
    getProfileLabel,
    CHECKPOINTS
};

File: backend/src/main/java/com/pocketcounselor/model/Checkpoint.java
package com.pocketcounselor.model;

import java.util.List;

public class Checkpoint {

    private final int level;
    private final int points;
    private final int minQuestions;
    private final String label;

    public static final List<Checkpoint> ALL = List.of(
            new Checkpoint(1, 280, 10, "Basic"),
            new Checkpoint(2, 420, 15, "Good"),
            new Checkpoint(3, 560, 20, "Very Detailed"),
            new Checkpoint(4, 700, 25, "Maximum")
    );

    public Checkpoint(int level, int points, int minQuestions, String label) {
        this.level = level;
        this.points = points;
        this.minQuestions = minQuestions;
        this.label = label;
    }

    public int getLevel() { return level; }
    public int getPoints() { return points; }
    public int getMinQuestions() { return minQuestions; }
    public String getLabel() { return label; }

    public static String getProfileLabel(int checkpointReached) {
        return ALL.stream()
                .filter(cp -> cp.level == checkpointReached)
                .map(Checkpoint::getLabel)
                .findFirst()
                .orElse("Preliminary");
    }
}
